prow/config/config.go

```diff
@@ -410,7 +410,7 @@ func yamlToConfig(path string, nc interface{}) error {
 	if err != nil {
 		return fmt.Errorf("error reading %s: %v", path, err)
 	}
-	if err := yaml.Unmarshal(b, nc); err != nil {
+	if err := yaml.UnmarshalStrict(b, nc); err != nil {
```

Comment:
Making this optional would be nice. This breaking change will be less painful if maintainers can address it when they are ready, or continue to use non-strict unmarshalling if they require that for some reason (not sure if there is a sane use case for that).

Reply:
imho there is no sane case for that, all of the things this is cactching are actual invalid configuration :grimacing:

Reply:
(I see the argument here and I thought about that, but on the other hand would you want prow to ignore unkown flags? configurations were incorrect / invalid if this change causes them to not parse)

Reply:
I think this is the wrong thing to do for prod

Reply:
Ideally we do strict for unit testing and loose for prod

Reply:
I don't think we should have more paths to loading config, I think we should have less :confused:

Reply:
Proto does not blow up when I send a message with unrecognized fields. We shouldn't do so here either. This is bad prod practice

Reply:
I agree with the intent and we should definitely ensure at unit test time that our config doesn't contain bad fields, but someone adding an extra field for whatever reason (processed by another utility, diagnostic info, etc) should no way blow up prow

Reply:
I'm prety sure ASCII protobuf used for config will. In either case, `kubectl create` | `kubectl apply` will not allow unknown fields either. I think I've only seen this considered bad practice for wire formats, not for configuration.

Reply:
note that prow will _not_ actually blow up, we catch panics, log them, and keep the current config

Reply:
That's why I'm thinking we prevent checking any config to git that has unknown fields.

On the other hand, what we upload into the cluster should be considered wire format and not fail in this way.

Reply:
>  but someone adding an extra field for whatever reason (processed by another utility, diagnostic info, etc) should no way blow up prow

that should absolutely be done via annotations.

Reply:
I don't think that is a wire format, what is in the configmap can even desync from git (and has) because of the way we do apply. When we fix that we typically don't interact with it from prow/ go code either. Other users may not use our unit tests but should have bogus config rejected to detect it early. 

To the same effect, if we did a direct `kubectl apply` isntead of the configmap updater we have, and these were actual k8s resource specs, it would be rejected with an error.

I've held and cancelled approval though since there seems to be non-zero agreement for both :upside_down_face:

Reply:
Even in cluster this is on-disk config & loading, rather than say, RPC calls. I respect the opposite viewpoint however. Will revisit this later.

Reply:
got back to this, working on a variant that allows supplying an unmarshal func and does not use strict for "prod-ish" loading. I still think we're more likely to catch issues with all-strict all-the-time however, but this version will be more flexible going forward.

Reply:
production should have been strict from the start.
now there are chances that prow users have bugs in their configs.

announcing 6 months ahead and then making a hard switch is also an option.